pkg/storage/swift/swift.go

```diff
@@ -1,84 +1,206 @@
+	if err != nil {
+		return nil, err
+	}
+
+	return client, nil
+}
+
+// NewDriver creates new Swift driver for the Image Registry
 func NewDriver(c *imageregistryv1.ImageRegistryConfigStorageSwift, listers *regopclient.Listers) *driver {
 	return &driver{
 		Config:  c,
 		Listers: listers,
 	}
 }
 
+// Secrets returns a map of the storage access secrets
 func (d *driver) Secrets() (map[string]string, error) {
+	cfg, err := clusterconfig.GetSwiftConfig(d.Listers)
+	if err != nil {
+		return nil, err
+	}
+
+	if cfg.Storage.Swift.Cloud == nil {
+		return map[string]string{
+			"REGISTRY_STORAGE_SWIFT_USERNAME": cfg.Storage.Swift.Username,
+			"REGISTRY_STORAGE_SWIFT_PASSWORD": cfg.Storage.Swift.Password,
+		}, nil
+	}
+
 	return nil, nil
 }
 
 func (d *driver) ConfigEnv() (envs []corev1.EnvVar, err error) {
+	// Add common values that do not depend on the config source
 	envs = append(envs,
 		corev1.EnvVar{Name: "REGISTRY_STORAGE", Value: "swift"},
-		corev1.EnvVar{Name: "REGISTRY_STORAGE_SWIFT_AUTHURL", Value: d.Config.AuthURL},
 		corev1.EnvVar{Name: "REGISTRY_STORAGE_SWIFT_CONTAINER", Value: d.Config.Container},
-		corev1.EnvVar{
-			Name: "REGISTRY_STORAGE_SWIFT_USERNAME",
-			ValueFrom: &corev1.EnvVarSource{
-				SecretKeyRef: &corev1.SecretKeySelector{
-					LocalObjectReference: corev1.LocalObjectReference{
-						Name: imageregistryv1.ImageRegistryPrivateConfiguration,
-					},
-					Key: "REGISTRY_STORAGE_SWIFT_USERNAME",
-				},
-			},
-		},
-		corev1.EnvVar{
-			Name: "REGISTRY_STORAGE_SWIFT_PASSWORD",
-			ValueFrom: &corev1.EnvVarSource{
-				SecretKeyRef: &corev1.SecretKeySelector{
-					LocalObjectReference: corev1.LocalObjectReference{
-						Name: imageregistryv1.ImageRegistryPrivateConfiguration,
-					},
-					Key: "REGISTRY_STORAGE_SWIFT_PASSWORD",
-				},
-			},
-		},
 	)
 
-	if d.Config.Domain != "" {
-		envs = append(envs, corev1.EnvVar{Name: "REGISTRY_STORAGE_SWIFT_DOMAIN", Value: d.Config.Domain})
+	cfg, err := clusterconfig.GetSwiftConfig(d.Listers)
+	if err != nil {
+		return nil, err
 	}
 
-	if d.Config.DomainID != "" {
-		envs = append(envs, corev1.EnvVar{Name: "REGISTRY_STORAGE_SWIFT_DOMAINID", Value: d.Config.DomainID})
+	if cfg.Storage.Swift.Cloud != nil {
+		cloud := cfg.Storage.Swift.Cloud
+		envs = append(envs,
+			corev1.EnvVar{Name: "REGISTRY_STORAGE_SWIFT_AUTHURL", Value: cloud.AuthInfo.AuthURL},
+			corev1.EnvVar{Name: "REGISTRY_STORAGE_SWIFT_DOMAIN", Value: cloud.AuthInfo.DomainName},
+			corev1.EnvVar{Name: "REGISTRY_STORAGE_SWIFT_DOMAINID", Value: cloud.AuthInfo.DomainID},
+			corev1.EnvVar{Name: "REGISTRY_STORAGE_SWIFT_TENANT", Value: cloud.AuthInfo.ProjectName},
+			corev1.EnvVar{Name: "REGISTRY_STORAGE_SWIFT_TENANTID", Value: cloud.AuthInfo.ProjectID},
+			corev1.EnvVar{Name: "REGISTRY_STORAGE_SWIFT_USERNAME", Value: cloud.AuthInfo.Username},
+			corev1.EnvVar{Name: "REGISTRY_STORAGE_SWIFT_PASSWORD", Value: cloud.AuthInfo.Password},
```

Comment:
You must not have secret information outside of the Kubernetes secrets. See https://github.com/openshift/cluster-image-registry-operator/blob/afd711132340f353f4f66e44e0949888b23324ac/pkg/storage/s3/s3.go#L108-L118

Reply:
This information comes from clouds.yaml, stored in "installer-cloud-credentials" secret. We read it with GetSwiftConfig function from clusterconfig.go

Reply:
These EnvVars go to the pod's spec.template.spec, i.e. they go to the deployment resource. So this code will expose passwords to users that have the cluster-reader role (they can't get secrets).